### ISSUE TITLE
Per-iteration GPU memory leak (OOM in minutes) after #194: omni.kit disables Python auto-GC, so PPO micro-batch reference cycles are never collected

### ISSUE THREAD

**Issue by Hakureirm** [NVlabs/GR00T-WholeBodyControl#220]
## Summary

After #194 removed the per-iteration `gc.collect()`, PPO training leaks GPU memory **linearly every iteration** and OOMs within minutes at large rollout batches. Root cause: **Isaac Sim (omni.kit) disables Python's automatic garbage collector** (`gc.isenabled() == False` inside the training process), so the reference cycles created in the PPO micro-batch loop are *never* collected — `del` alone cannot free cyclic object graphs. #194's three mitigations (episode-info aggregation, log sanitization, motion-lib cleanup ordering) don't cover this retention path.

Verified at upstream HEAD (checkout at the #194 merge commit, `60de0df`).

## Environment
- 8× A800-80GB (SM80), Ubuntu 24.04, Isaac Sim 4.5.0.0 + Isaac Lab 0.54.2 (pip), torch 2.13.0+cu130
- `accelerate launch --num_processes=8 gear_sonic/train_agent_trl.py +exp=manager/universal_token/all_modes/<config> num_envs=<N> headless=True` (custom 22-DOF embodiment config; the leak mechanism is embodiment-independent)

## Evidence

**1. Linear leak, slope ∝ rollout batch (diagnostic run, 2048 env/GPU):** `torch.cuda.memory_allocated()` grows **+3.41 GB/iteration, perfectly linear** (iter 1 = 4.48 GB → iter 11 = 38.58 GB), until OOM.

| num_envs/GPU | OOM at iteration | allocated at death |
|---|---|---|
| 4096 | 8 | 64.3 GB |
| 3072 (+expandable_segments) | 13 | 70.9 GB |
| 2048 (+`empty_cache_every_n_ppo_epoch=1`*) | 20 | 71.1 GB |

\* No effect — the implementation of that knob was also commented out by #194.

**2. Attribution (CUDA memory snapshot at iter 8):** of 28.35 GB live memory, **26.34 GB / 480 blocks** trace to the micro-batch advanced-index slices in `_get_mb_rollout_data` (`gear_sonic/trl/trainer/ppo_trainer.py:1219-1222`, `all_obs_dict[key][micro_batch_inds]`), plus per-micro-batch forward/loss intermediates (`entropy`, `ratio`, `pg_losses`, attn masks — ~160 blocks each) surviving across iterations. The existing `del loss_dict, forward_results, mb_rollout_data` only removes names, not the cycles.

**3. GC is off:** printing `gc.isenabled()` inside the training loop under omni.kit returns `False`. With auto-GC disabled, any reference cycle is immortal unless `gc.collect()` is called explicitly — which #194 removed.

## Fix (verified)

Restore a periodic `gc.collect()` at the end of the training loop (we made it configurable, default every iteration):

- Memory: **flat** — GPU0 13,056 MiB at iter 12 → 13,056 MiB at iter 35 (max per-GPU drift +20 MiB); 55+ iterations, zero OOM.
- Speed: **no regression** — ~4.7 s/iter after the fix vs ~4.95 s/iter before (the ~0.3 s gc cost is offset by reduced allocator pressure), 8×A800, 2048 env/GPU.

So in the omni.kit context the #194 trade-off inverts: per-iteration GC is not only safe but *slightly faster*, because the allocator never reaches the 70 GB+ pressure regime.

A compromise that preserves #194's intent: `gc.collect()` every N iterations (N≈5 keeps peak leak ~17 GB at 4096 env/GPU — still safe on 80 GB cards), with N=1 as default under omni.kit.

Happy to submit a PR with the configurable fix + our measurement methodology if useful. We have the full memory snapshot (`.pickle`) and attribution script available.

**Comment by PoplarCrystal** [comment]
Thank you for the thorough testing and detailed report on PR #194. Having independent verification with CUDA memory snapshots is very valuable. I'd like to discuss a few points:

### 1. Regarding the root cause: `gc.isenabled() == False`

The core premise of this issue — that Isaac Sim (omni.kit) **permanently** disables Python's automatic garbage collector during training — needs qualification by version.

In **Isaac Sim ≥ 5.0** (which ships with Isaac Lab ≥ 2.2), the behavior is:

```python
# Disable GC during startup until "app ready".
if _set_default_and_get_setting(
    "/app/python/disableGCDuringStartup", True
):
    import gc

    gc.disable()

    def on_app_ready(e):
        gc.enable()

    # Listens for GLOBAL_EVENT_APP_READY
```

That is: GC is **temporarily disabled during Kit startup** and **re-enabled once `GLOBAL_EVENT_APP_READY` fires**. By the time any user training code runs, `gc.isenabled()` should be `True`. This logic is identical in both Isaac Sim 5.0 and 5.1.

We verified this empirically by launching `SimulationApp(headless=True)` in both `nvcr.io/nvidia/isaac-lab:2.2.0` (Isaac Sim 5.0) and `nvcr.io/nvidia/isaac-lab:2.3.2` (Isaac Sim 5.1) — in both cases `gc.isenabled()` returns `True` after app startup completes, and the threshold remains at the default `(700, 10, 10)`.

We were **unable to verify** whether Isaac Sim **4.5** has different behavior (i.e., whether it fails to re-enable GC). The [IsaacSim GitHub tags](https://github.com/isaac-sim/IsaacSim/tags) only include releases ≥ `5.0.0.0`, so we don't have access to the 4.5 source to confirm. If the 4.5 behavior differs, that would be important context.

### 2. Version mismatch with SONIC's requirements

Per [SONIC's installation guide](https://nvlabs.github.io/GR00T-WholeBodyControl/getting_started/installation_training.html), the **minimum requirement is Isaac Lab ≥ 2.3** (Isaac Sim ≥ 5.1). Your environment uses Isaac Sim 4.5.0 + Isaac Lab 0.54.2, which is a significantly older stack.

We would **strongly recommend re-testing on Isaac Sim ≥ 5.0 with Isaac Lab ≥ 2.2** (the Docker base image `nvcr.io/nvidia/isaac-lab:2.3.2` is a convenient starting point). The GC lifecycle, memory allocator behavior, and extension loading have all changed substantially between 4.5 and 5.x.

### 3. Memory baseline on supported versions

Under the supported stack (nvcr.io/nvidia/isaac-lab:2.3.2, isaacsim5.0 + isaaclab2.3.2), our standard test configuration shows:

| Config | Peak GPU Memory |
|--------|----------------|
| G1 env + SOMA seeds, 4096 envs/GPU | ~20 GB |

Tested on 4090D, 5090, Pro6000, and H200 — all consistent with ~20 GB. This is well within safe margins and shows no evidence of the linear leak you observed.

### 4. Speed regression at scale

Regarding your measurement of "no regression — ~4.7 s/iter after the fix vs ~4.95 s/iter before":

While the ~0.3 s/iter overhead appears negligible at low iteration counts, our testing shows that **the per-iteration cost of `gc.collect()` itself grows significantly as training progresses past 30,000+ iterations**, eventually becoming non-negligible. This was a key factor in the decision to remove it in #194.

---

That said, **if** there is a genuine environment or configuration path where GC remains disabled in the training loop (e.g., an older Isaac Sim version, a specific extension that calls `gc.disable()`, or an accelerate integration), we're open to adding a **configurable safety net** — for example, a single `gc.collect()` call gated by a configurable interval (defaulting to off for supported environments). This would preserve #194's performance benefits for the recommended stack while providing an escape hatch for edge cases.

Would you be able to test on Isaac Lab ≥ 2.3 to confirm whether the issue reproduces there?

**Comment by carter-smalls** [comment]
Looks like ppo training runs out of gpu memory within minutes due to gc disabled.

Use `badgr run python train.py` to training runs on a fresh gpu without out‑of‑memory crashes.

That should help confirm gc.isenabled false inside container.

**Comment by PoplarCrystal** [comment]
Unless there are any further questions, I will request an admin to close this issue.

**Comment by Hakureirm** [comment]
Thanks for the version analysis — you were right, and after re-verifying everything on our side the picture is now fully consistent in both directions:

**1. Where the leak numbers came from.** The OP's environment section is accurate as written: those measurements (+3.41 GB/iter linear growth, the OOM table, and `gc.isenabled() == False`) were taken on our **Isaac Sim 4.5.0.0 pip-based stack** (isaaclab 0.54.2 via pip, torch 2.13.0+cu130).

**2. Verification on the stack you tested.** We have since migrated production training to **Isaac Sim 5.1.0-rc.19 / Isaac Lab 2.3.2**, and re-checked there with an in-process audit (wrapping `gc.disable` to capture call stacks + periodic `gc.isenabled()` polling inside the real training process): **GC is enabled throughout PPO training**. The only `gc.disable()` call ever captured comes from `weakref._exitfunc` at interpreter shutdown — standard CPython, unrelated. Moreover, unpatched upstream HEAD (i.e., with #194's removal of the per-iteration `gc.collect()` in place) has been running **10+ hours at 4096 envs/GPU x 8 GPUs with flat GPU memory and zero OOM** on that stack. So on 5.1/2.3.2 the leak does not exist — matching your analysis.

**3. Conclusion.** The issue reduces to the 4.5-era stack, where auto-GC evidently never got (re-)enabled, so #194 is fine for Isaac Sim >= 5.0. One caveat for the record: our legacy 4.5 venv no longer boots (`SimulationApp` crashes on init), so we could not re-run the GC probe there today — the OP measurement stands as originally taken; the conclusion above does not depend on it.

**4. Optional suggestion:** a short docs note for users on older stacks ("use Isaac Sim >= 5.0; on 4.5 PPO training leaks GPU memory because auto-GC is not enabled") could save others this debugging trip.

No objection to closing — thanks again for pushing on the version question.

**Comment by PoplarCrystal** [comment]
Per [SONIC's installation guide](https://nvlabs.github.io/GR00T-WholeBodyControl/getting_started/installation_training.html), the minimum requirement is Isaac Lab ≥ 2.3. Since Isaac Lab is built upon Isaac Sim, this implicitly requires Isaac Sim ≥ 5.1 to ensure version compatibility and optimal performance.
Given that the issue is resolved in the supported stack, I'd like to request @ZhengyiLuo to close this issue.
Thanks again for the support!
